### PR TITLE
[12.0][MIG] account_invoice_line_default_account

### PR DESCRIPTION
**Pull request by SirTakobi**
Migrating/Adding **account_invoice_line_default_account** to `12.0`.

This poor module has a troubled history: as far as I have found, it was brought from launchpad's `6.1` (unmerged) to `8.0` with https://github.com/OCA/account-invoicing/pull/23, then https://github.com/OCA/account-invoicing/pull/189, and finally https://github.com/OCA/account-invoicing/pull/356 (currently still open).

Then a migration to `10.0` happened in https://github.com/OCA/account-invoicing/pull/351 (currently still open).

I tried to put together something like a consistent commit history and this is the result, with a last commit for migrating to `12.0`.

Additional notes:

- I added some tests :tada:
- ~~There is a fixup commit because I couldn't find where the `__openerp__` was removed between https://github.com/OCA/account-invoicing/pull/356 and https://github.com/OCA/account-invoicing/pull/351, I didn't squash it in order not to mix authorships, but I can do it if @jbaudoux agrees~~
- About the roadmap:
  - For `account.invoice.line._onchange_account_id` I already tried substituting `write` with `update` but that results in not updating the partner (can be verified with the tests),
  - For `account.invoice.line.account_id`, currently the module assumes the previous `default` method is the one defined in **account** but that might not be True; retrieving the field's `default` inside the defined method is useless because it has already been changed to the defined method

  If you have any suggestion it is welcome

Looking at the linked PRs and commits, @jbaudoux you might want to have a look :)

### CODE REVIEW

File: account_invoice_line_default_account/models/account_invoice_line.py
Comment (jbaudoux) [inline account_invoice_line_default_account/models/account_invoice_line.py]: maybe better to use a different name to not have to call super

File: account_invoice_line_default_account/models/account_invoice_line.py
Comment (jbaudoux) [inline account_invoice_line_default_account/models/account_invoice_line.py]: self.env.context

**Review by jbaudoux — COMMENTED** [review]
Thanks for adding tests.

Some minor remarks.

I remember when I ported to v10 that lot of code had to be rewritten, that's why I squashed all the old commits. I don't think it is useful to recover all those old commits and I would squash them all into a single commit or at least reduce them a lot.

File: account_invoice_line_default_account/models/account_invoice_line.py
Comment (SirTakobi) [reply account_invoice_line_default_account/models/account_invoice_line.py]: Thanks for the review!
I can see no harm in calling `super`, actually it looks to me like a common thing to call `super` when overriding a method; why should this case be treated in a different way?

File: account_invoice_line_default_account/models/account_invoice_line.py
Comment (SirTakobi) [reply account_invoice_line_default_account/models/account_invoice_line.py]: Thanks, done :)

**Comment by SirTakobi** [comment]
> I remember when I ported to v10 that lot of code had to be rewritten, that's why I squashed all the old commits. I don't think it is useful to recover all those old commits and I would squash them all into a single commit or at least reduce them a lot.

@jbaudoux I squashed most of the commits, and added the **\_\_openerp__.py** removal in `10.0` migration, let me know if it looks better now.
Just as reference, I have kept the previous history in https://github.com/SirTakobi/account-invoicing/tree/12.0-mig-account_invoice_line_default_account-bkp

Can you update your review? Github is not allowing me to re-request you to do it

> /ocabot migration account_invoice_line_default_account

@rafaelbn thanks, I'm sorry I forgot to do those things

File: account_invoice_line_default_account/models/account_invoice_line.py
Comment (jbaudoux) [reply account_invoice_line_default_account/models/account_invoice_line.py]: Before `api.onchange`, there could be only a single method called on an onchange. With `api.onchange` you can have multiple different methods. The conditions are not to test if super has to be called or not but if this method has to be executed. So I find the current implementation hard to read.
Nevertheless it will work. Just a minor remark.

Comment (SirTakobi) [reply account_invoice_line_default_account/models/account_invoice_line.py]: That makes perfect sense: overriding the method is really needed only when you want to edit the method's behavior, but that does not have to be the way to go with the `onchange` methods.
I have done this change too, with a bit of explanation in the comments, let me know if it looks better now :)

Comment (jbaudoux) [reply account_invoice_line_default_account/models/account_invoice_line.py]: Looks clear now

File: account_invoice_line_sequence/__manifest__.py
Comment (jbaudoux) [inline account_invoice_line_sequence/__manifest__.py]: Why the 2 modules cannot work together?

Comment (SirTakobi) [reply account_invoice_line_sequence/__manifest__.py]: I wrote it in the commit message:
> This module is not compatible with account_invoice_line_sequence because them both add to the context of invoice lines in invoice form view

**Review by jbaudoux — APPROVED** [review]
LGTM

**Review by rafaelbn — APPROVED** [review]

File: .travis.yml
Comment (eLBati) [inline .travis.yml]: Can runboat be informed about this?

https://runboat.odoo-community.org/api/v1/builds/bf7e8fb88-cda7-4c02-a0c2-5717acb6d8bf/init-log

**Review by eLBati — CHANGES_REQUESTED** [review]
Proposing https://github.com/SirTakobi/account-invoicing/pull/1 to avoid basing on context and thus incompatibilities

File: .travis.yml
Comment (SirTakobi) [reply .travis.yml]: Runboat currently does not support conflicting addons, see https://github.com/sbidoul/runboat/issues/36

**Comment by SirTakobi** [comment]
> Proposing [SirTakobi#1](https://github.com/SirTakobi/account-invoicing/pull/1) to avoid basing on context and thus incompatibilities

@eLBati there are some things that do not really convince me in https://github.com/SirTakobi/account-invoicing/pull/1.

Instead of working on the linked PR, can we go on with the merge of this migration alone? Since we already have enough approvals, this module might finally be merged for the first time! :smile:
Then you will be able to propose that PR directly to `12.0`.
Unless you think there is something in this PR that might be blocking for the merge.

**Comment by SirTakobi** [comment]
@OCA/accounting-maintainers can this be merged or do I have to change something? Thanks!

**Comment by eLBati** [comment]
> Unless you think there is something in this PR that might be blocking for the merge

Uhm yes, this https://github.com/SirTakobi/account-invoicing/pull/1

**Comment by SirTakobi** [comment]
> > Unless you think there is something in this PR that might be blocking for the merge
> 
> Uhm yes, this [SirTakobi#1](https://github.com/SirTakobi/account-invoicing/pull/1)

Now that you explained those changes, I merged them :)

Just for future reference, my doubts were about choosing the onchange on `product_id` field specifically, but we clarified that that onchange is called on any line creation from UI; so the onchange can do the job that was previously (when onchanges were not called on line creation) performed from the line's view using the context.

**Comment by SirTakobi** [comment]
@eLBati can you update your review now? Thanks!

**Review by eLBati — APPROVED** [review]
Thanks

**Comment by SirTakobi** [comment]
@OCA/accounting-maintainers can this be merged? There are a few happy reviewers already :smile:

**Comment by SirTakobi** [comment]
@OCA/accounting-maintainers can this be merged? Thanks!

**Comment by SirTakobi** [comment]
@OCA/accounting-maintainers can this be merged? Thanks!

**Pull request merged**
